packages/tabsheet/src/vaadin-tabsheet.js

```diff
@@ -3,7 +3,269 @@
+        }
+
+        initCustomNode(tabs) {
+          if (!(tabs instanceof Tabs)) {
+            throw Error('The "tabs" slot of a <vaadin-tabsheet> must only contain a <vaadin-tabs> element!');
+          }
+          tabs.addEventListener('items-changed', tabsItemsChangedListener);
+          tabs.addEventListener('selected-changed', tabsSelectedChangedListener);
+          this.host.__tabs = tabs;
+        }
+
+        teardownNode(tabs) {
+          tabs.removeEventListener('items-changed', tabsItemsChangedListener);
+          tabs.removeEventListener('selected-changed', tabsSelectedChangedListener);
+          this.host._setItems([]);
+        }
+      })(this),
+    );
+
+    // Observe the panels slot for nodes. Set the assigned element nodes as the __panels array.
+    const panelSlot = this.shadowRoot.querySelector('#panel-slot');
+    this.__panelsObserver = new FlattenedNodesObserver(panelSlot, () => {
+      this.__panels = Array.from(panelSlot.assignedNodes({ flatten: true })).filter(
+        (node) => node.nodeType === Node.ELEMENT_NODE,
+      );
+    });
+  }
+
+  static get observers() {
+    return [
+      '__itemsOrPanelsChanged(items, __panels)',
+      '__selectedTabItemChanged(selected, items, __panels)',
+      '__propagateTabsProperties(__tabs, orientation, selected)',
+    ];
+  }
+
+  /**
+   * An observer which applies the necessary roles and ARIA attributes
+   * to associate the tab elements with the panels.
+   * @private
+   */
+  __itemsOrPanelsChanged(items, panels) {
+    if (!items || !panels) {
+      return;
+    }
+
+    items.forEach((tabItem) => {
+      const panel = panels.find((panel) => panel.getAttribute('tab') === tabItem.id);
+      if (panel) {
+        panel.role = 'tabpanel';
+        panel.id = panel.id || `tabsheet-panel-${generateUniqueId()}`;
+        panel.setAttribute('aria-labelledby', tabItem.id);
+
+        tabItem.setAttribute('aria-controls', panel.id);
+      }
+    });
+  }
+
+  /**
+   * An observer which toggles the visibility of the panels based on the selected tab.
+   * @private
+   */
+  __selectedTabItemChanged(selected, items, panels) {
+    if (!items || !panels || selected === undefined) {
+      return;
+    }
+
+    const selectedTab = items[selected];
+    const selectedTabId = selectedTab ? selectedTab.id : '';
+
+    panels.forEach((panel) => {
+      panel.hidden = panel.getAttribute('tab') !== selectedTabId;
+    });
+  }
+
+  /**
+   * An observer which propagates property value changes to the slotted <vaadin-tabs>.
+   * @private
+   */
+  __propagateTabsProperties(tabs, orientation, selected) {
+    if (tabs) {
+      tabs.orientation = orientation;
+      tabs.selected = selected;
+    }
+  }
+
+  /**
+   * An observer which reflects the orientation property to the host as aria-orientation.
+   * @private
+   */
+  __orientationChanged(orientation) {
+    if (orientation) {
+      this.setAttribute('aria-orientation', orientation);
```

Comment:
If we don't set `role="tablist"` then the `aria-orientation` can be also removed.

Reply:
Done